File: EVANS/frmMantVehiculo.vb
Imports System.Data.SqlClient

Public Class frmMantVehiculo

    Dim objVehiculo As New clsVehiculo
    Dim bolAccion As Boolean = True

    Private Sub frmMantVehiculo_FormClosing(ByVal sender As Object, ByVal e As System.Windows.Forms.FormClosingEventArgs) Handles Me.FormClosing

        Try

            LimpiarRAM()

        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub frmMantVehiculo_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

        Try
            Me.Top = 0
            Me.Left = 0

            TabControl1.TabPages(1).Enabled = False
            txtBuscar.Enabled = False

            btnGrabar.Enabled = False
            btnCancelar.Enabled = False
            btnEditar.Enabled = False
            btnBuscar.Enabled = False

            optBuscar.Checked = True

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub optTodos_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles optTodos.CheckedChanged

        Try
            txtBuscar.Clear()
            lvListado.Items.Clear()
            Dim drLista As SqlDataReader

            drLista = objVehiculo.Listar()

            While drLista.Read()
                Dim itemLista As ListViewItem
                itemLista = lvListado.Items.Add(NullToString(drLista("vehi_codigo")))
                itemLista.SubItems.Add(NullToString(drLista("vehi_marca")))
                itemLista.SubItems.Add(NullToString(drLista("vehi_placa")))
            End While
            drLista.Close()
            drLista.Dispose()

            txtBuscar.Enabled = False
            btnBuscar.Enabled = False

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub lvListado_DoubleClick(ByVal sender As Object, ByVal e As System.EventArgs) Handles lvListado.DoubleClick

        Try
            Dim strCodigo As String
            strCodigo = lvListado.SelectedItems.Item(0).SubItems(0).Text

            objVehiculo.Codigo = strCodigo
            If objVehiculo.BuscarXCodigo() Then

                TabControl1.TabPages(1).Enabled = True
                TabControl1.TabPages(0).Enabled = False
                TabControl1.SelectTab(1)

                txtCodigo.Text = objVehiculo.Codigo
                txtMarca.Text = objVehiculo.Marca
                txtPlaca.Text = objVehiculo.Placa
                txtConfiguracion.Text = objVehiculo.ConfiguracionVehicular
                txtCertificado.Text = objVehiculo.CertificadoInscripcion
                cbEmpresa.Text = objVehiculo.Empresa.Nombre
                cbEstado.Text = objVehiculo.Estado.Descripcion

                'Desactivar controles
                DesactivarControles(TabControl1, 1, True)

                'Comportamiento de botones
                btnNuevo.Enabled = True
                btnGrabar.Enabled = False
                btnEditar.Enabled = True
                btnCancelar.Enabled = True

                btnGrabar.Text = "Grabar"

            Else
                Return
            End If

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub btnNuevo_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnNuevo.Click

        Try
            objVehiculo = New clsVehiculo

            'Comportamiento del control TabControl
            TabControl1.TabPages(1).Enabled = True
            TabControl1.TabPages(0).Enabled = False
            TabControl1.SelectTab(1)
            txtMarca.Focus()

            'Limpiar y activa cajas de texto
            LimpiarControles(TabControl1, 1)
            DesactivarControles(TabControl1, 1, False)
            txtCodigo.ReadOnly = True

            'Comportamiento de botones
            btnNuevo.Enabled = False
            btnGrabar.Enabled = True
            btnCancelar.Enabled = True

            cbEstado.Text = "ACTIVO"

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub btnGrabar_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnGrabar.Click

        Try
            If txtMarca.Text = "" Or txtPlaca.Text = "" Or txtConfiguracion.Text = "" Or txtCertificado.Text = "" Or cbEmpresa.Text = "" Or cbEstado.Text = "" Then
                MessageBox.Show("Datos incompletos", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
                Return
            End If

            objVehiculo.Marca = txtMarca.Text.ToUpper
            objVehiculo.Placa = txtPlaca.Text.ToUpper
            objVehiculo.ConfiguracionVehicular = txtConfiguracion.Text.ToUpper
            objVehiculo.CertificadoInscripcion = txtCertificado.Text.ToUpper
            objVehiculo.Empresa.Codigo = BuscarCodigo("empr_codigo", "empr_nombre", cbEmpresa.Text, "Empresa")
            objVehiculo.Estado.Codigo = BuscarCodigo("esta_codigo", "esta_descripcion", cbEstado.Text, "Estado")

            If bolAccion = True Then
                If objVehiculo.Grabar() = False Then
                    Return
                End If
            Else
                If objVehiculo.Actualizar() = False Then
                    Return
                Else
                    bolAccion = True
                End If
            End If

            'Comportamiento del TabControl
            TabControl1.TabPages(0).Enabled = True
            TabControl1.TabPages(1).Enabled = False
            TabControl1.SelectTab(0)

            'Comportamiento de Botones
            btnNuevo.Enabled = True
            btnGrabar.Enabled = False
            btnEditar.Enabled = False
            btnCancelar.Enabled = False

            btnGrabar.Text = "Grabar"

            optBuscar.Checked = True

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub btnEditar_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnEditar.Click

        Try
            'Activa las cajas de texto
            DesactivarControles(TabControl1, 1, False)
            txtCodigo.ReadOnly = True
            txtMarca.Focus()

            'Comportamiento de botones
            btnNuevo.Enabled = False
            btnGrabar.Enabled = True
            btnEditar.Enabled = False
            btnCancelar.Enabled = True

            btnGrabar.Text = "Actualizar"
            bolAccion = False

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub btnCancelar_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnCancelar.Click

        Try
            'Comportamiento del TabControl
            TabControl1.TabPages(0).Enabled = True
            TabControl1.TabPages(1).Enabled = False
            TabControl1.SelectTab(0)

            bolAccion = True
            btnGrabar.Text = "Grabar"

            'Comportamiento de botones
            btnNuevo.Enabled = True
            btnGrabar.Enabled = False
            btnEditar.Enabled = False
            btnCancelar.Enabled = False

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub btnSalir_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnSalir.Click

        Try
            Me.Close()

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub btnBuscar_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles btnBuscar.Click

        Try
            If txtBuscar.Text = "" Then
                lvListado.Items.Clear()
                MessageBox.Show("Ingrese nombre a buscar", "Datos insuficientes", MessageBoxButtons.OK, MessageBoxIcon.Information)
                txtBuscar.Focus()
            Else
                lvListado.Items.Clear()
                Dim drLista As SqlDataReader
                drLista = Nothing

                If optMarca.Checked = True Then
                    objVehiculo.Marca = txtBuscar.Text
                    drLista = objVehiculo.BuscarXMarca()
                End If
                If optBuscar.Checked = True Then
                    objVehiculo.Placa = txtBuscar.Text
                    drLista = objVehiculo.BuscarXPlaca()
                End If

                While drLista.Read()
                    Dim itemLista As ListViewItem
                    itemLista = lvListado.Items.Add(NullToString(drLista("vehi_codigo")))
                    itemLista.SubItems.Add(NullToString(drLista("vehi_marca")))
                    itemLista.SubItems.Add(NullToString(drLista("vehi_placa")))
                End While
                drLista.Close()
            End If

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub optBuscar_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles optBuscar.CheckedChanged

        Try
            lvListado.Items.Clear()
            txtBuscar.Enabled = True
            txtBuscar.Focus()
            btnBuscar.Enabled = True

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub txtBuscar_KeyPress(ByVal sender As Object, ByVal e As System.Windows.Forms.KeyPressEventArgs) Handles txtBuscar.KeyPress

        Try
            If e.KeyChar = ChrW(Keys.Enter) Then
                e.Handled = True
                btnBuscar.PerformClick()
            End If

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub cbEmpresa_DropDown(ByVal sender As Object, ByVal e As System.EventArgs) Handles cbEmpresa.DropDown

        Try
            LLenarCombo("empr_codigo", "empr_nombre", "Empresa", cbEmpresa, True)

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub cbEstado_DropDown(ByVal sender As Object, ByVal e As System.EventArgs) Handles cbEstado.DropDown

        Try
            LLenarCombo("esta_codigo", "esta_descripcion", "Estado", cbEstado, False)

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub

    Private Sub optMarca_CheckedChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles optMarca.CheckedChanged

        Try
            lvListado.Items.Clear()
            txtBuscar.Enabled = True
            txtBuscar.Focus()
            btnBuscar.Enabled = True

        Catch ex As SqlException
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Catch ex As Exception
            MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            Return
        Finally
            If objConexion.State = ConnectionState.Open Then
                objConexion.Close()
            End If
        End Try

    End Sub
End Class

File: src/EVANS.UI.WinForms/Catalogo/frmMantVehiculo.cs
using System.Globalization;
using EVANS.Application.Catalogo.Commands;
using EVANS.Application.Catalogo.DTOs;
using EVANS.Application.Catalogo.Queries;
using MediatR;

namespace EVANS.UI.WinForms.Catalogo;

public partial class frmMantVehiculo : Form
{
    private readonly IMediator? _mediator;
    private IReadOnlyList<VehiculoDto> _vehiculos = [];
    private IReadOnlyList<EstadoDto> _estados = [];
    private IReadOnlyList<EmpresaDto> _empresas = [];
    private IReadOnlyList<EmpresaDto> _empresasActivas = [];
    private bool _isCreateMode = true;

    public frmMantVehiculo() => InitializeComponent();

    public frmMantVehiculo(IMediator mediator)
    {
        _mediator = mediator;
        InitializeComponent();
        WireEvents();
        ApplyInitialState();
    }

    internal int SelectedTabIndex => tabControl1.SelectedIndex;
    internal bool DetailsTabEnabled => tabPageDetalles.Enabled;
    internal bool ListingTabEnabled => tabPageListado.Enabled;
    internal bool BuscarEnabled => btnBuscar.Enabled;
    internal bool BuscarTextEnabled => txtBuscar.Enabled;
    internal bool NuevoEnabled => btnNuevo.Enabled;
    internal bool GrabarEnabled => btnGrabar.Enabled;
    internal bool EditarEnabled => btnEditar.Enabled;
    internal bool CancelarEnabled => btnCancelar.Enabled;
    internal string GrabarText => btnGrabar.Text;
    internal string CodigoText => txtCodigo.Text;
    internal string StatusMessage => lblmsg.Text;
    internal int ListCount => lvListado.Items.Count;
    internal string? FirstListMarca => lvListado.Items.Count == 0 ? null : lvListado.Items[0].SubItems[1].Text;
    internal string? FirstListPlaca => lvListado.Items.Count == 0 ? null : lvListado.Items[0].SubItems[2].Text;
    internal void SetTestSearchText(string value) => txtBuscar.Text = value;
    internal void SetTestMarca(string value) => txtMarca.Text = value;
    internal void SetTestPlaca(string value) => txtPlaca.Text = value;
    internal void SetTestConfiguracion(string value) => txtConfiguracion.Text = value;
    internal void SetTestCertificado(string value) => txtCertificado.Text = value;
    internal void SetTestEmpresa(int codigo) => SelectEmpresa(codigo);
    internal void SetTestEstado(int codigo) => SelectEstado(codigo);
    internal void SelectAllSearchForTest() => optTodos.Checked = true;
    internal void SelectMarcaSearchForTest() => optMarca.Checked = true;
    internal void SelectPlacaSearchForTest() => optBuscar.Checked = true;
    internal void BeginNewForTest() => btnNuevo_Click();
    internal void BeginEditForTest() => btnEditar_Click();

    internal async Task OpenFirstForTestAsync()
    {
        if (lvListado.Items.Count > 0)
            await OpenVehiculoAsync(int.Parse(lvListado.Items[0].SubItems[0].Text, CultureInfo.InvariantCulture));
    }

    internal async Task LoadVehiculosAsync()
    {
        await EnsureCombosAsync();
        _vehiculos = await Mediator.Send(new ListVehiculosMaintenanceQuery());
        BindVehiculos(_vehiculos);
    }

    internal async Task<bool> BuscarAsync(bool showMessages = false)
    {
        if (string.IsNullOrWhiteSpace(txtBuscar.Text))
        {
            lvListado.Items.Clear();
            SetStatus("Ingrese nombre a buscar");
            if (showMessages)
                MessageBox.Show("Ingrese nombre a buscar", "Datos insuficientes", MessageBoxButtons.OK, MessageBoxIcon.Information);
            txtBuscar.Focus();
            return false;
        }

        if (_vehiculos.Count == 0)
            _vehiculos = await Mediator.Send(new ListVehiculosMaintenanceQuery());

        var search = txtBuscar.Text.Trim();
        var filtered = optMarca.Checked
            ? _vehiculos.Where(v => (v.Marca ?? string.Empty).StartsWith(search, StringComparison.CurrentCultureIgnoreCase))
            : _vehiculos.Where(v => v.Placa.StartsWith(search, StringComparison.CurrentCultureIgnoreCase));
        BindVehiculos(filtered);
        SetStatus("Los campos marcados con asterisco (*) son obligatorios");
        return true;
    }

    internal async Task<bool> SaveAsync(bool showMessages = false)
    {
        var empresaCodigo = ResolveEmpresaCodigo();
        var estadoCodigo = ResolveEstadoCodigo();
        if (string.IsNullOrWhiteSpace(txtMarca.Text) || string.IsNullOrWhiteSpace(txtPlaca.Text)
            || string.IsNullOrWhiteSpace(txtConfiguracion.Text) || string.IsNullOrWhiteSpace(txtCertificado.Text)
            || empresaCodigo is null || estadoCodigo is null)
        {
            SetStatus("Datos incompletos");
            if (showMessages)
                MessageBox.Show("Datos incompletos", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
            return false;
        }

        var marca = txtMarca.Text.Trim().ToUpperInvariant();
        var placa = txtPlaca.Text.Trim().ToUpperInvariant();
        var configuracion = txtConfiguracion.Text.Trim().ToUpperInvariant();
        var certificado = txtCertificado.Text.Trim().ToUpperInvariant();

        if (_isCreateMode)
        {
            var result = await Mediator.Send(new CreateVehiculoCommand(
                marca, placa, configuracion, certificado, empresaCodigo.Value, estadoCodigo.Value));
            if (!result.IsSuccess) return ShowFailure(result.Error, showMessages);
        }
        else
        {
            var result = await Mediator.Send(new UpdateVehiculoCommand(
                int.Parse(txtCodigo.Text, CultureInfo.InvariantCulture), marca, placa, configuracion, certificado,
                empresaCodigo.Value, estadoCodigo.Value));
            if (!result.IsSuccess) return ShowFailure(result.Error, showMessages);
            _isCreateMode = true;
        }

        await ReturnToListingAsync();
        return true;
    }

    private void WireEvents()
    {
        Load += async (_, _) => await EnsureCombosAsync();
        optTodos.CheckedChanged += async (_, _) => await optTodos_CheckedChanged();
        optMarca.CheckedChanged += (_, _) => SearchOption_CheckedChanged(optMarca);
        optBuscar.CheckedChanged += (_, _) => SearchOption_CheckedChanged(optBuscar);
        lvListado.DoubleClick += async (_, _) => await lvListado_DoubleClick();
        btnNuevo.Click += (_, _) => btnNuevo_Click();
        btnGrabar.Click += async (_, _) => await SaveAsync(showMessages: true);
        btnEditar.Click += (_, _) => btnEditar_Click();
        btnCancelar.Click += async (_, _) => await ReturnToListingAsync();
        btnSalir.Click += (_, _) => Close();
        btnBuscar.Click += async (_, _) => await BuscarAsync(showMessages: true);
        txtBuscar.KeyPress += async (_, e) =>
        {
            if (e.KeyChar == (char)Keys.Enter)
            {
                e.Handled = true;
                await BuscarAsync(showMessages: true);
            }
        };
        cbEstado.DropDown += async (_, _) => await EnsureEstadosAsync();
        cbEmpresa.DropDown += async (_, _) => await EnsureEmpresasAsync();
    }

    private void ApplyInitialState()
    {
        Top = 0;
        Left = 0;
        tabPageDetalles.Enabled = false;
        txtBuscar.Enabled = false;
        btnGrabar.Enabled = false;
        btnCancelar.Enabled = false;
        btnEditar.Enabled = false;
        btnBuscar.Enabled = false;
        optBuscar.Checked = true;
    }

    private async Task optTodos_CheckedChanged()
    {
        if (!optTodos.Checked) return;
        txtBuscar.Clear();
        await EnsureEmpresasAsync();
        if (_vehiculos.Count == 0)
            _vehiculos = await Mediator.Send(new ListVehiculosMaintenanceQuery());
        BindVehiculos(_vehiculos);
        txtBuscar.Enabled = false;
        btnBuscar.Enabled = false;
    }

    private void SearchOption_CheckedChanged(RadioButton option)
    {
        if (!option.Checked) return;
        lvListado.Items.Clear();
        txtBuscar.Enabled = true;
        txtBuscar.Focus();
        btnBuscar.Enabled = true;
    }

    private async Task lvListado_DoubleClick()
    {
        if (lvListado.SelectedItems.Count > 0)
            await OpenVehiculoAsync(int.Parse(lvListado.SelectedItems[0].SubItems[0].Text, CultureInfo.InvariantCulture));
    }

    private async Task OpenVehiculoAsync(int codigo)
    {
        await EnsureCombosAsync();
        var vehiculo = await Mediator.Send(new GetVehiculoByIdQuery(codigo));
        if (vehiculo is null) return;

        tabPageDetalles.Enabled = true;
        tabPageListado.Enabled = false;
        tabControl1.SelectTab(tabPageDetalles);
        txtCodigo.Text = vehiculo.Codigo.ToString(CultureInfo.InvariantCulture);
        txtMarca.Text = vehiculo.Marca;
        txtPlaca.Text = vehiculo.Placa;
        txtConfiguracion.Text = vehiculo.ConfiguracionVehicular;
        txtCertificado.Text = vehiculo.CertificadoInscripcion;
        SelectEmpresa(vehiculo.EmpresaCodigo);
        SelectEstado(vehiculo.EstadoCodigo);
        SetDetailInputsEnabled(false);
        btnNuevo.Enabled = true;
        btnGrabar.Enabled = false;
        btnEditar.Enabled = true;
        btnCancelar.Enabled = true;
        btnGrabar.Text = "Grabar";
    }

    private void btnNuevo_Click()
    {
        _isCreateMode = true;
        tabPageDetalles.Enabled = true;
        tabPageListado.Enabled = false;
        tabControl1.SelectTab(tabPageDetalles);
        txtCodigo.Clear();
        txtMarca.Clear();
        txtPlaca.Clear();
        txtConfiguracion.Clear();
        txtCertificado.Clear();
        cbEmpresa.SelectedIndex = -1;
        cbEstado.SelectedIndex = -1;
        SetDetailInputsEnabled(true);
        txtCodigo.ReadOnly = true;
        txtMarca.Focus();
        btnNuevo.Enabled = false;
        btnGrabar.Enabled = true;
        btnCancelar.Enabled = true;
        btnEditar.Enabled = false;
        btnGrabar.Text = "Grabar";
    }

    private void btnEditar_Click()
    {
        SetDetailInputsEnabled(true);
        txtCodigo.ReadOnly = true;
        txtMarca.Focus();
        btnNuevo.Enabled = false;
        btnGrabar.Enabled = true;
        btnEditar.Enabled = false;
        btnCancelar.Enabled = true;
        btnGrabar.Text = "Actualizar";
        _isCreateMode = false;
    }

    private async Task ReturnToListingAsync()
    {
        tabPageListado.Enabled = true;
        tabPageDetalles.Enabled = false;
        tabControl1.SelectTab(tabPageListado);
        _isCreateMode = true;
        btnGrabar.Text = "Grabar";
        btnNuevo.Enabled = true;
        btnGrabar.Enabled = false;
        btnEditar.Enabled = false;
        btnCancelar.Enabled = false;
        optBuscar.Checked = true;
        _vehiculos = await Mediator.Send(new ListVehiculosMaintenanceQuery());
        lvListado.Items.Clear();
    }

    private async Task EnsureCombosAsync()
    {
        await EnsureEstadosAsync();
        await EnsureEmpresasAsync();
    }

    private async Task EnsureEstadosAsync()
    {
        if (_estados.Count > 0) return;
        _estados = await Mediator.Send(new ListEstadosQuery());
        cbEstado.DataSource = _estados.ToList();
        cbEstado.DisplayMember = nameof(EstadoDto.Descripcion);
        cbEstado.ValueMember = nameof(EstadoDto.Codigo);
    }

    private async Task EnsureEmpresasAsync()
    {
        if (_empresas.Count == 0)
            _empresas = await Mediator.Send(new ListEmpresasMaintenanceQuery());
        if (_empresasActivas.Count > 0) return;
        _empresasActivas = await Mediator.Send(new ListEmpresasQuery());
        cbEmpresa.DataSource = _empresasActivas.ToList();
        cbEmpresa.DisplayMember = nameof(EmpresaDto.RazonSocial);
        cbEmpresa.ValueMember = nameof(EmpresaDto.Codigo);
    }

    private void SelectEmpresa(int codigo)
    {
        if (_empresasActivas.Count == 0) return;
        cbEmpresa.SelectedValue = codigo;
        if (!Equals(cbEmpresa.SelectedValue, codigo)) cbEmpresa.Text = GetEmpresaName(codigo);
    }

    private void SelectEstado(int codigo)
    {
        if (_estados.Count > 0) cbEstado.SelectedValue = codigo;
    }

    private int? ResolveEmpresaCodigo()
    {
        if (cbEmpresa.SelectedValue is not null)
            return Convert.ToInt32(cbEmpresa.SelectedValue, CultureInfo.InvariantCulture);
        return _empresas.FirstOrDefault(e => string.Equals(e.RazonSocial, cbEmpresa.Text, StringComparison.CurrentCultureIgnoreCase))?.Codigo;
    }

    private int? ResolveEstadoCodigo()
    {
        if (cbEstado.SelectedValue is not null)
            return Convert.ToInt32(cbEstado.SelectedValue, CultureInfo.InvariantCulture);
        return _estados.FirstOrDefault(e => string.Equals(e.Descripcion, cbEstado.Text, StringComparison.CurrentCultureIgnoreCase))?.Codigo;
    }

    private void SetDetailInputsEnabled(bool enabled)
    {
        txtCodigo.Enabled = enabled;
        txtMarca.Enabled = enabled;
        txtPlaca.Enabled = enabled;
        txtConfiguracion.Enabled = enabled;
        txtCertificado.Enabled = enabled;
        cbEmpresa.Enabled = enabled;
        cbEstado.Enabled = enabled;
    }

    private void BindVehiculos(IEnumerable<VehiculoDto> vehiculos)
    {
        lvListado.Items.Clear();
        foreach (var vehiculo in vehiculos)
        {
            var item = lvListado.Items.Add(vehiculo.Codigo.ToString(CultureInfo.InvariantCulture));
            item.SubItems.Add(vehiculo.Marca ?? string.Empty);
            item.SubItems.Add(vehiculo.Placa);
        }
    }

    private string GetEmpresaName(int codigo) =>
        _empresas.FirstOrDefault(e => e.Codigo == codigo)?.RazonSocial
        ?? _empresasActivas.FirstOrDefault(e => e.Codigo == codigo)?.RazonSocial
        ?? string.Empty;

    private bool ShowFailure(string? error, bool showMessages)
    {
        var message = error ?? "No se pudo guardar el vehículo.";
        SetStatus(message);
        if (showMessages) MessageBox.Show(message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
        return false;
    }

    private void SetStatus(string message) => lblmsg.Text = message;

    private IMediator Mediator => _mediator
        ?? throw new InvalidOperationException("Use the IMediator constructor at runtime. The parameterless constructor is for the WinForms designer only.");
}
